java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java

```diff
@@ -0,0 +1,152 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.adapter.jdbc;
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.Iterator;
+
+import org.apache.arrow.adapter.jdbc.consumer.CompositeJdbcConsumer;
+import org.apache.arrow.adapter.jdbc.consumer.JdbcConsumer;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+/**
+ * VectorSchemaRoot iterator for partially converting JDBC data.
+ */
+public class ArrowVectorIterator implements Iterator<VectorSchemaRoot> {
+
+  private final ResultSet resultSet;
+  private final JdbcToArrowConfig config;
+
+  private final Schema schema;
+  private final ResultSetMetaData rsmd;
+
+  private final JdbcConsumer[] consumers;
+  private final CompositeJdbcConsumer compositeConsumer;
+  private boolean consumerCreated = false;
+
+  private VectorSchemaRoot nextBatch;
+
+  private final int partialLimit;
+
+  /**
+   * Construct an instance.
+   */
+  private ArrowVectorIterator(ResultSet resultSet, JdbcToArrowConfig config) throws SQLException {
+    this.resultSet = resultSet;
+    this.config = config;
+    this.schema = JdbcToArrowUtils.jdbcToArrowSchema(resultSet.getMetaData(), config);
+    this.partialLimit = config.getPartialLimit();
+
+    rsmd = resultSet.getMetaData();
+    consumers = new JdbcConsumer[rsmd.getColumnCount()];
+    this.compositeConsumer = new CompositeJdbcConsumer(consumers);
+  }
+
+  /**
+   * Create a ArrowVectorIterator to partially convert data.
+   */
+  public static ArrowVectorIterator create(
+      ResultSet resultSet,
+      JdbcToArrowConfig config)
+      throws SQLException {
+
+    ArrowVectorIterator iterator = new ArrowVectorIterator(resultSet, config);
+    try {
+      iterator.load();
+      return iterator;
+    } catch (Exception e) {
+      iterator.close();
+      throw new RuntimeException("error occurs while creating iterator", e);
+    }
+  }
+
+  // Loads the next schema root or null if no more rows are available.
+  private void load() throws IOException, SQLException {
+    VectorSchemaRoot root = VectorSchemaRoot.create(schema, config.getAllocator());
+    JdbcToArrowUtils.allocateVectors(root, JdbcToArrowUtils.DEFAULT_BUFFER_SIZE);
+
+    if (!consumerCreated) {
```

Comment:
can this be moved to the constructor? that would eliminate the state variable?

Reply:
I guess not, now consumer is created with schema root, and constructor dosen't has root yet?

Reply:
> I guess not, now consumer is created with schema root, and constructor dosen't has root yet?

Also I think without change the consumer creation logic, this flag is not proper to remove. Otherwise we may need to create a schema root in constructor specific for creating consumers which is not really used in follow up.